### PR TITLE
relax mul! type constraints

### PR DESCRIPTION
**Pull request by JeffFessler**
i hope this solves #40

### CODE REVIEW

**Comment by dlfivefifty** [comment]
Can you add a test as well

**Comment by JeffFessler** [comment]
Unfortunately my first attempt here is failing the existing tests and I won't have time for a while to dig further..  Yes, it needs further tests after it passes the existing ones.  Help wanted!

**Comment by JeffFessler** [comment]
@dlfivefifty I have fixed the bugs in my initial PR and added tests that address the issues raised with #40 and now it passes the CI tests.

There were a couple methods that purported to be there to avoid method ambiguities, but they seemed to be unnecessary (and causing problems) with this new eltype flexibility.  So I block commented them out.  But I did not delete them in case they are helpful later for some corner case the current tests have overlooked.  I also ran `detect_ambiguities(ToeplitzMatrices)` and it did not find any issues.

Now, this PR does not solve everything.  We can now create a Toeplitz matrix from integer arguments, but because of the FFTs involved the product with anything (including a vector of Ints) will still be Float.   Someone else can tackle that in another PR.  At least here we can now do the product and get some results.

Lots of junk commits here so please squash when the time comes...

File: src/ToeplitzMatrices.jl [hunk 1]
```diff
@@ -43,7 +43,10 @@ end
 convert(::Type{Matrix}, A::AbstractToeplitz) = Matrix(A)
 
 # Fast application of a general Toeplitz matrix to a column vector via FFT
-function mul!(y::StridedVector{T}, A::AbstractToeplitz{T}, x::StridedVector, α::T, β::T) where T
+function mul!(y::StridedVector, A::AbstractToeplitz, x::StridedVector, α::Number, β::Number)
+    T = promote_type(eltype.([y, A, x, α, β])...)
+#   toreal = x -> (T <: Real) ? convert(T, x) : x # causes InexactError
+    toreal = x -> (T <: Real) ? real(x) : x
```
Comment (dlfivefifty) [inline src/ToeplitzMatrices.jl:49]: What is the point of `toreal`?  Why make this an anonymous function? The previous version is simpler

File: src/ToeplitzMatrices.jl [hunk 2]
```diff
@@ -108,9 +113,11 @@ function mul!(C::StridedMatrix{T}, A::AbstractToeplitz{T}, B::StridedMatrix, α:
     return C
 end
 # Avoid ambiguity error
+#=
```
Comment (dlfivefifty) [inline src/ToeplitzMatrices.jl:116]: Just delete code if it's not needed unless you expect it to be needed in the future: git history will save it around in case

File: src/ToeplitzMatrices.jl (hunk 1 above)
Comment (JeffFessler) [reply src/ToeplitzMatrices.jl]: Ok, I will use the ternary instead.  Might be better for the compiler I guess, and branch prediction will be perfect.

**Comment by JeffFessler** [comment]
I've added some docstrings now, and I also generalized a couple constructors so that they can take an AbstractVector because `Circulant(1:3)` is convenient.  And I've added tests for that generality.

File: src/ToeplitzMatrices.jl (hunk 2 above)
Comment (JeffFessler) [reply src/ToeplitzMatrices.jl]: Done.

**Comment by dlfivefifty** [comment]
Great, thanks! Would you mind adding codecov as part of this PR? Then I can merge and tag.

**Comment by JeffFessler** [comment]
Ok, I cut and paste some lines from config files in FillArrays and we'll see if I did it correctly.
But I think you will have to enable codecov for this project on Travis yourself.

**Comment by dlfivefifty** [comment]
Thanks. It should work automatically...

**Comment by JeffFessler** [comment]
In case you didn't notice, I bumped the version so you may need to trigger the registration.  Thanks.

**Comment by dlfivefifty** [comment]
I already did, 13 minutes ago....
